src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs

```diff
@@ -1708,6 +1763,61 @@ private static unsafe string Int32ToHexStr(int value, char hexBase, int digits)
             return buffer;
         }
 
+        private static unsafe string UInt32ToBinaryStr(uint value, int digits)
```

Comment:
Any reason these aren't a shared generic implementation?

Reply:
I was just copying what was done for hex.  I can change it, though.  What are you envisioning for a shared generic implementation?

Reply:
There are three implementations: one for UInt128, one for UInt64, and one for UInt32.  The UInt128 one just delegates once or twice to the UInt64 one.  The UInt64 one has two implementations; on 32-bit it just delegates once or twice to the UInt32 one.  The 64-bit implementation of UInt64 and UInt32 are both the loop cited.  So we could save one such method in the case of 64-bit, at the expense of a bit more complicated ifdef'ing.  Not sure it's worth it.

I do in general like the idea of consolidating the formatting half as you did with the parsing half.  Let's take another whack at it after things settle.